### PR TITLE
feat(deepseek_v4): FP8 batched BMM for grouped output LoRA wo_a

### PR DESCRIPTION
**Pull request by zufayu**
## Summary

Replace the BF16 grouped einsum at the end of `DeepseekV4Attention.forward` with `aiter.ops.triton.gemm.batched.batched_gemm_a8w8_a_per_token_group_prequant_w_per_batched_tensor_quant` (same Triton kernel MLA's `_v_up_proj_and_o_proj` already uses).

Builds on PR #650 (DeepSeek V4 PR1 skeleton). Touches one file: `atom/models/deepseek_v4.py`.

## What changed

- `__init__`: read `ATOM_V4_OA_USE_EINSUM` env once into `self._use_oa_einsum`.
- `process_weights_after_loading`:
  - **Default (env=0, BMM path)**: dequant on-disk FP8+per-128-block-scale wo_a → bf16 → reshape `(G_local, R, K)` → `dynamic_per_batched_tensor_quant` (inlined as `_v4_dynamic_per_batched_tensor_quant`) → store as `self.W_OA` (FP8 buffer) + `self.W_OA_scale` (scalar buffer). Replace `self.wo_a.weight` with 1-element placeholder. Adds order assertion + idempotent guard.
  - **env=1 (einsum fallback)**: existing dequant path, no requant.
- forward (the 4 lines):
  - Default: `_aiter_triton_fp8_bmm(o, W_OA, W_OA_scale, group_size=128, transpose_bm_in=True, transpose_bm=True)` — single fused launch with on-the-fly per-token-group X quant; output layout matches einsum, so `wo_b` reshape unchanged.
  - env=1: legacy `torch.einsum("sgd,grd->sgr", o, wo_a)`.

## Verification (PR #650 standard test config)

```bash
ATOM_USE_TRITON_MOE=1 AITER_LOG_LEVEL=WARNING \
python -m atom.examples.simple_inference \
  --model /path/to/DeepSeek-V4-Pro \
  --kv_cache_dtype fp8 -tp 8 \
  --max-num-seqs 4 --max-num-batched-tokens 1024 --max-model-len 1024 \
  --gpu-memory-utilization 0.85 --enforce-eager \
  --temperature 0.0 --max-tokens 128
```

### Correctness (4-prompt batch on V4-Pro / 8x MI355X)

All 4 prompts produce coherent, semantically equivalent output to einsum baseline. Deterministic prompts match by content; open-ended prompts diverge in expected FP8 quant noise pattern (early edge-confidence token flips, then content develops independently — both sides remain coherent and factually correct):

| Prompt | BMM | einsum | Comment |
|---|---|---|---|
| `1+2+3=?` | `= **6**` (16 tok eos) | `= 6` (15 tok eos) | bold token diff (logit perturbation) |
| `list primes < 100` | identical | identical | high-confidence tokens stable |
| `introduce yourself` | "...Chinese company DeepSeek..." | "...DeepSeek (深度求索) company..." | both correct intros |
| `如何增肌10公斤` | "几乎不可能实现的目标" | "在生理上几乎不可能实现的目标" | both correct advice |

### Perf (eager mode, ATOM_USE_TRITON_MOE=1, --max-tokens 128, V4-Pro on 8x MI355X)

Same docker session, all caches warm, **apples-to-apples**:

| Req | output | einsum | **BMM** | Δ |
|---|---|---|---|---|
| Req 2 (1+2+3) | 16 tok eos | 0.514s | 0.518s | +0.8% |
| Req 1 (primes) | 94 tok eos | 0.455s | 0.449s | **−1.3%** |
| Req 0 (intro) | 128 tok max | 0.440s | 0.433s | **−1.6%** |
| Req 3 (增肌) | 128 tok max | 0.440s | 0.433s | **−1.6%** |
| **TTFT** (post first-call cache) | — | 2.99s | 3.31s | +10% noise |

**TPOT: parity (−1.6 ~ +0.8%, all within run-to-run noise).**

### Honest kernel-level finding (see RFC comment below)

Single-op micro-bench reveals the FP8 BMM kernel itself is **1.5-2x slower** than rocBLAS BF16 BMM (what `torch.einsum` lowers to) at V4 wo_a's shape (B=2, K=4096, N=1024). TPOT shows parity only because wo_a is < 0.5% of total V4 forward time. Tuning aiter's JSON config DB does not close the gap — the bottleneck is fundamental (existing batched kernels grid over B, designed for MLA's B≈128).

The RFC comment on this PR proposes a new aiter Triton kernel (`batched_gemm_a8w8_smallB_blockscale`) that would address this regime properly. **This PR is a placeholder**: it captures the production code change and validates correctness/memory; it should land alongside or after the new aiter kernel for a real perf win.

### Memory

VRAM saved: ~2-3% per rank (88% vs 90-91% on V4-Pro 8x MI355X), measured during `simple_inference` run after warmup. Direct contribution: wo_a runtime storage drops from BF16 ~16MB/layer to FP8 ~8MB/layer + scalar scale (~256MB/rank total; remainder of measured delta likely from KV cache adaptive sizing).

## Trade-off

| Dimension | Status |
|---|---|
| Correctness (PR #650 smoke prompts) | Pass |
| TPOT (eager, V4 PR std config) | Parity (−1.6 ~ +0.8%) |
| TTFT (steady-state) | Parity |
| wo_a kernel time (single-op) | **Slower 1.5-2x vs einsum** (see RFC) |
| Memory (per rank) | **−2~3% headroom** |
| FP8 precision impact | Per-block-scale → scalar-scale collapse; lm_eval not yet measured (test node was offline) |

## Toggle

`ATOM_V4_OA_USE_EINSUM=1` falls back to legacy BF16 einsum path. Read once at `__init__`, zero per-forward overhead. Useful for A/B comparison and debug rollback.

## Open question for reviewer

Holding this PR open while the RFC for a new aiter kernel (in the comment thread) is evaluated. Two paths forward:

1. **Merge as-is** — accept memory-only win, kernel-level perf regression invisible at TPOT level. Useful as a staging step so V4 isn't blocked on aiter changes.
2. **Wait for new aiter kernel, then refactor this PR** — `process_weights_after_loading` becomes a no-op (LinearBase standard FP8 path), forward calls the new kernel directly. Larger eventual win (perf parity / improvement + memory + zero precision loss).

Defer to V4 PR #650 maintainer + aiter team.

## File touched

`atom/models/deepseek_v4.py` (+145, −47)

### CODE REVIEW

**Comment by zufayu** [comment]
# RFC: New aiter Triton kernel for small-B FP8 batched GEMM

## Motivation

DeepSeek V4's grouped-output-LoRA `wo_a` projection has shape (per TP=8 rank):
- `B = n_local_groups = 2` (the kernel batch dimension)
- `K = d_per_group = 4096`
- `N = o_lora_rank = 1024`
- `M = token batch` (1-16 in decode, up to ~1024 in chunked prefill)

This shape is materially different from MLA's regime where the existing aiter
batched FP8 kernels were tuned (B = num_heads ≈ 128). At B=2 the existing
kernels lose to PyTorch's BF16 grouped einsum (which lowers to rocBLAS BMM).

### Measured (V4-Pro, gfx950, single-op micro-bench)

| M | einsum (BF16) | `batched_gemm_a8w8_a_per_token_group_prequant_w_per_batched_tensor_quant` (FP8) | G-loop `gemm_a8w8_blockscale_preshuffle` (FP8) |
|---|---|---|---|
| 1 | 15 μs | 35 μs (**0.43x**) | 116 μs (**0.13x**) |
| 16 | 14 μs | 35 μs (**0.43x**) | 60 μs (**0.24x**) |
| 64 | 19 μs | 37 μs (**0.51x**) | 62 μs (**0.30x**) |
| 256 | 19 μs | 36 μs (**0.54x**) | 58 μs (**0.33x**) |
| 1024 | 37 μs | 39 μs (0.96x) | 55 μs (0.67x) |

(Reproducible via `scripts/v4_wo_a_tune.py --microbench` and
 `scripts/v4_wo_a_microbench_gloop.py` on the PR #676 branch.)

The FP8 paths are **1.5x – 7x slower** than the BF16 baseline at this shape,
even though FP8 should theoretically halve the W bandwidth requirement.

## Why existing kernels lose

### `batched_gemm_a8w8_a_per_token_group_prequant_w_per_batched_tensor_quant`
Grid topology is `(B, M_tiles * N_tiles)`. At B=2, the first grid dimension
provides only 2-way parallelism. With BLOCK_SIZE_N=128 (the default for this
shape) and N=1024 → 8 N-tiles. Decode M=4 → 1 M-tile. **Total work-groups: 16.
On 304 CU MI355X this is ~5% utilization.** Tuning BLOCK sizes can raise
work-group count but introduces atomic / launch overhead that cancels the gain
(verified empirically with both `_get_config` JSON tuning and `@triton.autotune`
sweep approaches).

### G-loop `gemm_a8w8_blockscale_preshuffle`
Standard 2D GEMM per group, no batching. Each call does fill the chip well, but
runs G_local=2 times serially with G+1 kernel launches per layer (one act_quant
+ G GEMMs). Eager-mode launch overhead dominates at small M; CUDAGraph would
help but isn't yet available on V4 (per #650's "Known limitations").

### rocBLAS / hipBLAS BMM (what `torch.einsum` lowers to)
Wins decisively at B=2 small-M because:
- Hand-tuned vendor kernel for small-batch BMM
- Mature autotune database covering exactly this shape regime
- Single launch, full-CU coverage even at B=2 via internal split-K / persistent
  scheduling

## Proposed: `batched_gemm_a8w8_smallB_blockscale`

A new aiter Triton kernel optimized for the `small B + medium-large K + medium N`
regime, using **on-disk per-128-block W scale directly** (no requant, no
precision loss).

### Key design choices

1. **Collapse `B` into the grid's M dimension.** Instead of grid `(B, M*N/tile)`,
   use a 1D grid `(B*ceil(M/BM)*ceil(N/BN),)` with internal program-id
   decomposition. Lets the launcher dispatch all `B*M*N` work-groups across CUs
   in one wave, avoiding the B=2 bottleneck.

2. **Mandatory split-K.** K=4096 with M small is bandwidth-bound; split-K=4 or 8
   adds a 3rd grid dimension and a small reduction kernel, giving 4-8x more
   work-groups.  This is the standard fix for "tall-skinny" GEMMs and is already
   in `gemm_a8w8_blockscale.py` (just not in any batched kernel).

3. **Native per-block W scale support** (vs requant-to-scalar in
   `batched_gemm_a8w8_a_per_token_group_prequant_w_per_batched_tensor_quant`).
   Eliminates load-time precision collapse. Implementation: load `w_scale[b,
   n_block, k_block]` per inner-loop iteration; `BLOCK_SIZE_K = 128` is already
   constrained by per-token-group quant on the X side.

4. **Preserve fused per-token-group X act-quant** (same as today's
   `_a_per_token_group_prequant_*` kernel). On-the-fly quant of X removes the
   need for a separate act-quant kernel call from the caller.

### Signature sketch

```python
batched_gemm_a8w8_smallB_blockscale_a_per_token_group_prequant_w_per_block_scale(
    X,               # (B, M, K) BF16 — quantized to FP8 inline
    WQ,              # (B, N, K) FP8 — pre-quantized at load
    w_scale,         # (B, N//128, K//128) FP32 — per-128-block W scale
    group_size=128,  # X act-quant block size on K (= W block size on K)
    block_n=128,     # W block size on N
    transpose_bm=False,
    transpose_bm_in=False,
    splitK=4,        # NEW: required for small-B / small-M regimes
    dtype=torch.bfloat16,
)
```

### Expected benefits

- **wo_a kernel parity or beat vs einsum** — collapsing B into grid + split-K
  should restore CU utilization that the existing batched kernel sacrifices.
- **Eliminate one-time precision loss** of #676's per-block→scalar requant.
- **No load-time dequant + requant** dance — `process_weights_after_loading`
  becomes a no-op (LinearBase's standard FP8 + shuffle handles wo_a like any
  other Linear).
- **Reusable beyond V4** — any future model with grouped output projections
  (small B, medium-large K) benefits.

### Implementation effort

- Triton kernel: ~150 LoC (start from
  `batched_gemm_a8w8_a_per_token_group_prequant_w_per_batched_tensor_quant.py`,
  swap scalar w_scale load for per-block load, add 1D grid + split-K +
  reduction kernel).
- Wrapper + autotune config: ~50 LoC.
- Unit tests: ~80 LoC (compare vs reference einsum).
- Estimated 1-2 engineer-days for a working version, +1 day for autotune
  baseline.

## Alternative we considered and rejected

- **`@triton.autotune` patch on existing kernel** — sweeps configs, picks
  best per shape. Tried; even with full sweep + low-noise bench, no config
  beats the default by > 5%. The B=2 grid bottleneck is fundamental, not a
  config choice. Tuning improves wo_a kernel by < 10% (still 1.4x slower
  than einsum at small M).

- **`split_K` retrofit on existing kernel** — could land in the same place
  as the new kernel's split-K. Worth exploring as an incremental option,
  but the existing kernel's `(B, M*N/tile)` grid still wastes the B
  dimension; cleanest fix is the new kernel.

- **Continue with PR #676 as-is** — kernel-level 2x slower than einsum;
  TPOT shows parity only because wo_a is < 0.5% of forward time. Memory
  savings of −2~3% per rank are the only real win and may not justify
  the added complexity.

## Ask

1. Aiter team: is there appetite to land this kernel? Happy to draft the
   PR if there's a path to merge.
2. V4 PR #650 maintainer: would this kernel close PR #676 (BMM) decisively
   in favor of the new path, or does the current memory-only trade still
   warrant landing PR #676 in the interim?

## Reproduce

Bench scripts on PR #676 branch (`feat/deepseek-v4-wo-a-fp8-bmm`):
```bash
python scripts/v4_wo_a_tune.py --microbench           # einsum vs current FP8 BMM
python scripts/v4_wo_a_microbench_gloop.py            # einsum vs G-loop blockscale
```

**Comment by zufayu** [comment]
Tracked aiter side at ROCm/aiter#3000 — proposing a new `batched_gemm_a8w8_smallB_blockscale_a_per_token_group_prequant_w_per_block_scale` kernel optimized for V4 wo_a's `(B=2, K=4096, N=1024)` shape. This PR will refactor (process_weights_after_loading → no-op, forward → call new kernel directly) once that kernel lands.

**Comment by zufayu** [comment]
Superseded by #773. Original branch was 62 commits behind main; clean re-application from current main HEAD with newly-documented secondary win (BF16 direct_copy elimination via transpose_bm=True).
